src/diagnostics.rs

```diff
@@ -0,0 +1,118 @@
+use crate::term::{
+  check::{
+    ctrs_arities::ArityErr, exhaustiveness::ExhaustivenessErr, set_entrypoint::EntryErr,
+    shared_names::TopLevelErr, type_check::InferErr, unbound_pats::UnboundCtr, unbound_vars::UnboundVar,
+  },
+  transform::{
+    extract_adt_matches::MatchError, resolve_refs::ReferencedMain, simplify_ref_to_ref::CyclicDef,
+  },
+  Name,
+};
+use itertools::Itertools;
+use std::fmt::{Display, Formatter};
+
+#[derive(Debug, Clone, Default)]
+pub struct Info {
+  err_counter: usize,
+  pub errs: Vec<Error>,
+  pub warns: Vec<Warning>,
+}
+
+impl Info {
+  pub fn error<E: Into<Error>>(&mut self, e: E) {
+    self.errs.push(e.into())
+  }
+
+  /// Updates the internal counter to the current number of errors
+  pub fn start_pass(&mut self) {
+    self.err_counter = self.errs.len();
+  }
+
+  /// Checks if any error was emitted since the start of the pass,  
+  /// Returning all the current errors as a `Err(String)`.  
+  /// Otherwise, returns the given arg as an `Ok(T)`.
+  pub fn fatal<T>(&mut self, t: T) -> Result<T, Info> {
+    if self.errs.len() == self.err_counter { Ok(t) } else { Err(std::mem::take(self)) }
+  }
+}
+
+impl Display for Info {
+  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+    write!(f, "{}", self.errs.iter().join("\n"))
+  }
+}
+
+impl From<String> for Info {
+  fn from(value: String) -> Self {
+    Info { err_counter: 0, errs: vec![Error::Custom(value)], warns: Vec::new() }
+  }
+}
+
+impl From<&str> for Info {
+  fn from(value: &str) -> Self {
+    Info::from(value.to_string())
+  }
+}
+
+// TODO: Merge errors that reference the same definition
+#[derive(Debug, Clone)]
+pub enum Error {
+  Exhaustiveness(Name, ExhaustivenessErr),
+  MainRef(Name, ReferencedMain),
```

Comment:
Can you add `Err` to all these error type names?
`ReferencedMainErr`, `UnboundVarErr`, etc.
When a function returns `ReferencedMain` it's a bit confusing without context